### PR TITLE
Interesting Crafting

### LINKED ISSUE #13380: Materials.
We have a material datum system but it's more some counters for abstractly storing material amounts of a set list of minerals than actually having meaningful materials.

On bay and goon and such on the other hand, a lot more things have materials incorporated. You don't just have a table, you have a steel table, etc... This is really cool and it opens up a lot of flexibility and personally I think materials are a better way to handle upgrades than flat part upgrades.

On goon you can literally combine Uqill, a very blast resistant crystal, with Fibrilith, a sort of cloth-like "mineral", To make blast resistant cloth. Now make a jumpsuit out of that. You can make rainbow coloured floor tiles. And jumpsuits. And chairs. And toolboxes. You can make plasma chairs which produce plasma gas when sparked. Hide some in the bushes and run around spamming RCD sparks for !!FUN!!

The problem is our materials system isn't exactly laid out to handle this. At all.

Now the question is, how do we make a system like this here?

### PR DESCRIPTION
**Pull request by MadmanMartian**
closes #13380

- [x] Inherit color from base material type
- [ ] Make use of some form of functionality for base material type
- [x] Add quality handling

:cl: 
* rscadd: Adds material inheritance to created objects

### CODE REVIEW

File: __DEFINES/obj_defines.dm [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#define NORMAL 3
+#define GOOD 4
+#define SUPERIOR 5
+#define EXCELLENT 6
+#define MASTERWORK 7
+#define LEGENDARY 8
+
+/proc/getQualityString(var/quality)
+	switch(quality)
+		if(0)
```
Comment (ghost) [inline __DEFINES/obj_defines.dm]: you literally defined the defines above you nerd.
Also in my opinion I'd make it an assoc list ie
quality_to_string = list("[AWFUL]" ="awful","[SHODDY]" = "shoddy") etc so you could just return quality_to_string[quality]

File: __DEFINES/obj_defines.dm [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#define SHODDY 1
+#define POOR 2
+#define NORMAL 3
+#define GOOD 4
+#define SUPERIOR 5
+#define EXCELLENT 6
+#define MASTERWORK 7
+#define LEGENDARY 8
+
+/proc/getQualityString(var/quality)
```
Comment (ghost) [inline __DEFINES/obj_defines.dm]: no camel case >:(

File: code/game/objects/items/stacks/stack_recipes.dm [hunk 3]
```diff
@@ -69,6 +69,45 @@
...
+	var/datum/material/mat
+	var/datum/materials/materials_list = new
+	if(istype(S, /obj/item/stack/sheet/mineral)) //With mineral, it's straight forward
+		var/obj/item/stack/sheet/mineral/M = S
+
+		mat = materials_list.getMaterial(M.recyck_mat)
+	else //Not so straight forward
+		switch(S.type)
+			if(/obj/item/stack/sheet/metal)
+				mat = materials_list.getMaterial(MAT_IRON)
```
Comment (ghost) [inline code/game/objects/items/stacks/stack_recipes.dm]: There's gotta be a better way to do this.

File: code/game/objects/items/stacks/stack_recipes.dm [hunk 4]
```diff
@@ -69,6 +69,45 @@
...
+			if(/obj/item/stack/sheet/wood)
+				mat = materials_list.getMaterial(MAT_WOOD)
+		if(istype(S, /obj/item/stack/sheet/glass))
+			mat = materials_list.getMaterial(MAT_GLASS)
+	to_chat(usr, "mat = [mat], inherit_material = [inherit_material], gen_quality = [gen_quality]")
```
Comment (ghost) [inline code/game/objects/items/stacks/stack_recipes.dm]: uh oh

File: code/game/objects/items/stacks/stack_recipes.dm [hunk 5]
```diff
@@ -69,6 +69,45 @@
...
+	if(mat)
+		if(inherit_material)
+			to_chat(world, "inherit_material called by [mat]")
```
Comment (ghost) [inline code/game/objects/items/stacks/stack_recipes.dm]: uh oh

File: code/game/objects/items/stacks/stack_recipes.dm [hunk 6]
```diff
@@ -69,6 +69,45 @@
...
+			var/icon/original = icon(R.icon, R.icon_state)
+			original.ColorTone(mat.color)
+			R.icon = original
+			R.alpha = mat.alpha
+			R.material_type = mat
+			to_chat(world, "material type of R is [R.material_type]")
```
Comment (ghost) [inline code/game/objects/items/stacks/stack_recipes.dm]: UH OH

File: code/game/objects/objs.dm [hunk 7]
```diff
@@ -31,7 +31,9 @@ var/global/list/reagents_to_log = list(FUEL, PLASMA, PACID, SACID, AMUTATIONTOXI
 	plane = OBJ_PLANE
 
 	var/defective = 0
-
+	var/quality = NORMAL //What leve of quality this object is.
```
Comment (ghost) [inline code/game/objects/objs.dm]: LEVE

File: code/modules/mining/materials.dm [hunk 8]
```diff
@@ -127,11 +127,16 @@ var/global/list/initial_materials	//Stores all the matids = 0 in helping New
 	var/sheettype=null
 	var/cointype=null
 	var/value=0
+	var/color
```
Comment (ghost) [inline code/modules/mining/materials.dm:130]: aren't you a br*t?
image

**Review by ghost — CHANGES_REQUESTED** [review]

File: code/modules/mining/materials.dm (hunk 8 above)
Comment (MadmanMartian) [reply code/modules/mining/materials.dm:130]: I have not had enough sleep, will fix all but this, as codebase standard is english simplified, not english traditional

Comment (ghost) [reply code/modules/mining/materials.dm:130]: further proof that the brits ruin everything

Comment (SonixApache) [reply code/modules/mining/materials.dm:130]: color is fine

Comment (Probe1) [reply code/modules/mining/materials.dm:130]: >English (simplified)

Ow

**Comment by carshalash** [comment]
Can I get a masterwork plasma sword?

**Comment by ghost** [comment]
This is a ☼Railgun☼ .

will items found on the astroid with dwarf fortress descriptions be effected in any way?

**Comment by MadmanMartian** [comment]
no, but that is a good idea on where to start on making the description changes, should the thing created by a masterpiece

**Comment by MadmanMartian** [comment]
@carshalash yes

**Comment by MadmanMartian** [comment]
training_swords
More PoC things, training swords in this case.

Blending looks okay, gotta work on my shading though.

**Comment by Probe1** [comment]
Yeah there's not much outline on them.

**Comment by MadmanMartian** [comment]
happy carp

It's a start

File: code/modules/mining/materials.dm [hunk 9]
```diff
@@ -127,11 +127,24 @@ var/global/list/initial_materials	//Stores all the matids = 0 in helping New
...
 
 /datum/material/New()
 	if(processed_name=="")
 		processed_name=name
 
+/datum/material/proc/on_use(obj/source, atom/target, mob/user)
+	ASSERT(source)
+	if(isobserver(user))
+		return
+	to_chat(world, "[src].on_use called, source is [source], target is [target], user is [user]")
```
Comment (MadmanMartian) [inline code/modules/mining/materials.dm]: HITLERS

**Comment by MadmanMartian** [comment]
> Iamgoofball - Today at 3:54 AM
make the quality a list and just pull quality with indexing

**Comment by Zth--** [comment]
what determines the quality of the item? the material? would a skill system like dorf work in /vg/?

**Comment by ghost** [comment]
@MadmanMartian THAT'S WHAT I SAID WAY BEFORE GOOF SAID IT RREEE

**Comment by MadmanMartian** [comment]
@Zth-- semi-randomness, the material, and the surroundings

**Comment by jknpj** [comment]
AAAAAAAAAAAAAAAAAAA

**Comment by MadmanMartian** [comment]
mother FUCK

**Comment by FudgePucker** [comment]
Silver swords should do more damage to vampires

**Comment by MadmanMartian** [comment]
@FudgePucker that is a very good idea, but would rather get the concept in, then go for potential balance changes

File: code/game/objects/items/stacks/tiles/tile_types.dm [hunk 10]
```diff
@@ -102,7 +102,8 @@
 	if(iswrench(W))
 		if(use(4))
 			playsound(src, 'sound/items/Ratchet.ogg', 50, 1)
-			drop_stack(/obj/item/stack/sheet/wood, get_turf(user), 1, user)
+			playsound(get_turf(src), 'sound/items/Ratchet.ogg', 50, 1)
```
Comment (MadmanMartian) [inline code/game/objects/items/stacks/tiles/tile_types.dm]: Whoops

File: code/modules/mining/materials.dm (hunk 8 above)
Comment (nuklearcellphoneg) [reply code/modules/mining/materials.dm:130]: AMERICANA GODDAMNIT

**Comment by MadmanMartian** [comment]
globalaccess is a pain

**Comment by FudgePucker** [comment]
What about instead of adding new ores, you add a cruciable to mining outpost? It could take existing ores and give you a chance to purify it into a smaller amount of a new metal.
Maybe even add metal alloying to it, letting you craft up base metals into new ones.

New machines to craft stuff, a normal lathe that would let you make gun barrels out of tungsten. INB4 wyci

File: code/modules/mining/materials.dm [hunk 11]
```diff
@@ -149,6 +166,18 @@ var/global/list/initial_materials	//Stores all the matids = 0 in helping New
 	cc_per_sheet=CC_PER_SHEET_GLASS
 	oretype=/obj/item/weapon/ore/glass
 	sheettype=/obj/item/stack/sheet/glass/glass
+	color = "#6E8DA2" //rgb: 110, 141, 162
+	alpha = 122
+	brunt_damage_mod = 0.7
+	sharpness_mod = 1.4
+
+/datum/material/glass/on_use(obj/source)
+	..()
```
Comment (DamianX) [inline code/modules/mining/materials.dm]: shouldn't the parent return something and shouldn't you check that something? as things are right now, calling the parent is useless

**Review by DamianX — DISMISSED** [review]

**Comment by chesse20** [comment]
ADD THIS PLEASE

**Comment by MadmanMartian** [comment]
there's literally only the two things on it. Need more sprites for objects!¬

File: code/game/objects/objs.dm [hunk 12]
```diff
@@ -615,3 +631,41 @@ a {
...
+/obj/proc/gen_description(mob/user)
+	var/material_mod = quality-GOOD>1 ? quality-GOOD : 0
+	var/additional_description
+	if(material_mod)
+		additional_description = "On \the [src] is a carving, it depicts:\n"
+		var/list/characters = list()
+		for(var/i = 1 to material_mod)
+			if(prob(50)) //We're gonna use an atom
+				var/atom/AM = pick(existing_typesof(/mob/living/simple_animal))
+				if(!characters.Find(initial(AM.name)))
```
Comment (DamianX) [inline code/game/objects/objs.dm]: replace these two lines with `characters |= initial(AM.name)`

File: code/game/objects/objs.dm [hunk 13]
```diff
@@ -615,3 +631,41 @@ a {
...
+					characters.Add(initial(AM.name))
+			else
+				var/strangething = pick("captain","clown","mime","\improper CMO","cargo technician","medical doctor","[user ? user : "stranger"]","octopus","changeling","\improper Nuclear Operative", "[pick("greyshirt", "greytide", "assistant")]", "xenomorph","catbeast","[user && user.mind && user.mind.heard_before.len ? pick(user.mind.heard_before) : "strange thing"]","Central Command","\improper Ian","[ticker.Bible_deity_name]","Nar-Sie","\improper Poly the Parrot","\improper Wizard","vox")
+				if(!characters.Find(strangething))
```
Comment (DamianX) [inline code/game/objects/objs.dm]: replace these two lines with `characters |= strangething`

File: code/game/objects/objs.dm [hunk 14]
```diff
@@ -615,3 +631,41 @@ a {
...
+					characters.Add(strangething)
+			additional_description += "[i == material_mod ? " & a " : "[i > 1 ? ", a ": " A "]"][characters[i]]"
+		additional_description += ". They are in \the [pick("captains office","Space","mining outpost","vox outpost","a space station","[station_name()]","bar","kitchen","library","Science","void","Bluespace","Hell","Central Command")]"
+		if(material_mod > 1)
```
Comment (DamianX) [inline code/game/objects/objs.dm]: If I'm reading this correctly this can never be false at this point

File: code/game/objects/objs.dm [hunk 15]
```diff
@@ -615,3 +631,41 @@ a {
...
+			additional_description += ". They are [pick("[pick("fighting","robusting","attacking","beating up", "abusing")] [pick("each other", pick(characters))]","playing cards","firing lasers at [pick("something",pick(characters))]","crying","laughing","blank faced","screaming","cooking [pick("something", pick(characters))]", "eating [pick("something", pick(characters))]")]. "
+		if(characters.len)
```
Comment (DamianX) [inline code/game/objects/objs.dm]: superfluous check

File: code/game/objects/objs.dm [hunk 16]
```diff
@@ -59,7 +67,11 @@ var/global/list/reagents_to_log = list(FUEL, PLASMA, PACID, SACID, AMUTATIONTOXI
 	if(integratedpai)
 		qdel(integratedpai)
 		integratedpai = null
+	if(on_use)
+		on_use.holder = null
+	qdel(on_use)
```
Comment (DamianX) [inline code/game/objects/objs.dm]: indent this under the if and null it afterwards

**Comment by optimumtact** [comment]
when did goof start coding for vg?

**Comment by SonixApache** [comment]
long time ago actually

**Pull request merged**